gpiozero/pins/native.py

```diff
@@ -98,52 +100,173 @@ class GPIOFS(object):
 
-    def __init__(self):
-        self._lock = Lock()
-        self._pin_refs = Counter()
+    def __init__(self, factory, queue):
+        self._lock = RLock()
+        self._exports = {}
+        self._thread = NativeWatchThread(factory, queue)
+
+    def close(self):
+        # We *could* track the stuff we've exported and unexport it here, but
+        # exports are a system global resource. We can't guarantee that some
+        # other process isn't relying on something we've exported. In other
+        # words, once exported it's *never* safe to unexport something. The
+        # unexport method below is largely provided for debugging and testing.
+        if self._thread is not None:
+            self._thread.close()
+            self._thread = None
 
     def path(self, name):
         return os.path.join(self.GPIO_PATH, name)
 
+    def path_value(self, pin):
+        return self.path('gpio%d/value' % pin)
+
+    def path_dir(self, pin):
+        return self.path('gpio%d/direction' % pin)
+
+    def path_edge(self, pin):
+        return self.path('gpio%d/edge' % pin)
+
+    def exported(self, pin):
+        return pin in self._exports
+
     def export(self, pin):
         with self._lock:
-            if self._pin_refs[pin] == 0:
-                # Set the count to 1 to indicate the GPIO is already exported
-                # (we'll correct this if we find it isn't, but this enables us
-                # to "leave the system in the state we found it")
-                self._pin_refs[pin] = 1
+            try:
+                result = self._exports[pin]
+            except KeyError:
                 result = None
                 # Dirty hack to wait for udev to set permissions on
-                # gpioN/direction; there's no other way around this as there's
-                # no synchronous mechanism for setting permissions on sysfs
+                # gpioN/value; there's no other way around this as there's no
+                # synchronous mechanism for setting permissions on sysfs
                 for i in range(10):
                     try:
-                        result = io.open(self.path('gpio%d/value' % pin), 'w+b', buffering=0)
+                        # Must be O_NONBLOCK for use with epoll in edge
+                        # triggered mode
+                        result = os.open(self.path_value(pin),
+                                         os.O_RDONLY | os.O_NONBLOCK)
                     except IOError as e:
                         if e.errno == errno.ENOENT:
-                            with io.open(self.path('export'), 'wb') as f:
+                            with io.open(self.path('export'), 'w+b') as f:
                                 f.write(str(pin).encode('ascii'))
-                            # Pin wasn't exported, so correct the ref-count
-                            self._pin_refs[pin] = 0
                         elif e.errno == errno.EACCES:
                             sleep(i / 100)
```

Comment:
Should there be some kind of timeout code here (and also below), to prevent it sleep()-ing forever if/when something goes wrong?